adoc/extensions/sycl_khr_free_function_commands.adoc

```diff
@@ -0,0 +1,994 @@
+
+[source,c++]
+----
+h.set_args(args...);
+h.parallel_for(r, k);
+----
+
+_Effects (4-6)_: Equivalent to [code]#q.submit([&](handler& h) { launch(h, r, k,
+args...); })#.
+
+'''
+
+.[apititle]#launch_reduce# (kernel function)
+[source,role=synopsis,id=api:launch_reduce]
+----
+namespace sycl::khr {
+
+template <typename KernelType, typename... Reductions>
+void launch_reduce(sycl::handler& h, sycl::range<1> r,
+                   const KernelType& k, Reductions&&... reductions); (1)
+
+template <typename KernelType, typename... Reductions>
+void launch_reduce(sycl::handler& h, sycl::range<2> r,
+                   const KernelType& k, Reductions&&... reductions); (2)
+
+template <typename KernelType, typename... Reductions>
+void launch_reduce(sycl::handler& h, sycl::range<3> r,
+                   const KernelType& k, Reductions&&... reductions); (3)
+
+template <typename KernelType, typename... Reductions>
+void launch_reduce(sycl::queue q, sycl::range<1> r,
+                   const KernelType& k, Reductions&&... reductions); (4)
+
+template <typename KernelType, typename... Reductions>
+void launch_reduce(sycl::queue q, sycl::range<2> r,
+                   const KernelType& k, Reductions&&... reductions); (5)
+
+template <typename KernelType, typename... Reductions>
+void launch_reduce(sycl::queue q, sycl::range<3> r,
+                   const KernelType& k, Reductions&&... reductions); (6)
+
+}
+----
+_Constraints_: The parameter pack consists of 1 or more objects created by the
+[code]#sycl::reduction# function.
+
+_Effects (1-3)_: Equivalent to [code]#+h.parallel_for(r, reductions..., k)+#.
+
+_Effects (4-6)_: Equivalent to [code]#+q.submit([&](handler& h) {
+launch_reduce(h, r, k, reductions...); })+#.
+
+
+'''
+
+.[apititle]#launch_reduce# (kernel object)
+[source,role=synopsis,id=api:launch_reduce-kernel]
+----
+namespace sycl::khr {
+
+template <typename... Args>
+void launch_reduce(sycl::handler& h, sycl::range<1> r,
+                   const sycl::kernel& k, Args&&... args); (1)
+
+template <typename... Args>
+void launch_reduce(sycl::handler& h, sycl::range<2> r,
+                   const sycl::kernel& k, Args&&... args); (2)
+
+template <typename... Args>
+void launch_reduce(sycl::handler& h, sycl::range<3> r,
+                   const sycl::kernel& k, Args&&... args); (3)
+
+template <typename... Args>
+void launch_reduce(sycl::queue q, sycl::range<1> r,
+                   const sycl::kernel& k, Args&&... args); (4)
+
+template <typename... Args>
+void launch_reduce(sycl::queue q, sycl::range<2> r,
+                   const sycl::kernel& k, Args&&... args); (5)
+
+template <typename... Args>
+void launch_reduce(sycl::queue q, sycl::range<3> r,
+                   const sycl::kernel& k, Args&&... args); (6)
+
+}
+----
+_Effects (1-3)_: Equivalent to:
+[source,c++]
+----
+h.set_args(args...);
+h.parallel_for(r, k);
+----
+
+_Effects (4-6)_: Equivalent to [code]#q.submit([&](handler& h) {
+launch_reduce(h, r, k, args...); })#.
```

Comment:
`launch_reduce` performs a reduction and `launch` does not.

I don't know exactly how to specify this, though. My understanding is that the order in which arguments are passed to `sycl::kernel` launches is implementation-defined.  Do you think it would be sufficient to say something like:

> _Preconditions_:
> * The kernel performs 1 or more reductions; and
> * `Args` contains the required backend-specific arguments for reduction kernels.

Reply:
We haven't defined any backend-specific arguments that correspond to the "reduction" object.  Therefore, I don't think it's possible to do backend interop with a "reduction" object.  The other use for `kernel` objects is the free-function kernel extension (an Intel extension), but free-function kernels do not support "reduction" objects either.  It's not clear to me that there is currently any use for this overload of `launch_reduce`.

Reply:
I agree, but because the missing information is backend-specific, I think we need to define this overload in the specification.  Otherwise, there's no way for a backend to specify how to call a `kernel` object associated with a reduction. 

Any compiled kernel with a `reducer` argument knows what reduction is being performed, and `launch_reduce` knows how to launch that kernel and initialize all the data.  The only thing missing is what a user should pass to `set_args`, and I believe it's possible to define that (for a specific backend).

Reply:
> Otherwise, there's no way for a backend to specify how to call a `kernel` object associated with a reduction.

In the core spec, a `kernel` object can represent either:

* A device bulit-in kernel, or
* A kernel created from backend interop, for example creating a kernel from an OpenCL `cl_kernel`.

Therefore, the form of `launch` that takes a `kernel` is basically providing a way to launch a backend-specific kernel from SYCL.  A backend-specific kernel will never have an argument that is a "reduction" object because the reduction object is a SYCL-specific concept.

If that doesn't convince you, here is another argument.  Even if there was some way to create a `kernel` object that took a "reduction", the kernel would certainly have other non-reduction parameters.  Therefore, some of the arguments would be "reductions" and others would not.  I don't see why we would want a special function name for the case when some arguments are reductions.  We would instead expand the set of legal arguments for `handler::set_arg` in general, and then users could use the regular `launch` overload that takes a `kernel` object.

Reply:
I don't think it's this simple.  Reductions are somewhat of a special case in SYCL, because there are quite a few implementations where a reduction is actually executed via multiple kernels.

In essence, you can view SYCL's reduction syntax as a shorthand for defining and launching a much more complicated sequence of kernels.  A SYCL kernel _never_ takes a `sycl::reduction` as an argument (note that the object is passed as an argument to `parallel_for`,  rather than being captured by the lambda), and by the time the user's lambda is converted into a kernel, it's been wrapped a few times.

The "real" reduction kernel often looks something like this:

```c++
template <typename UserProvidedLambda, typename Reduction>
struct RealReductionKernel {

  // This is the kernel function.
  void operator()(nd_item it) {
  
    // Initialize the reducer(s).
    reducer r = ...;
    
    // Launch the user-provided function.
    f(it, r);
    
    // Contribute results from the reducer(s) back to memory.
    *some_memory[it.get_group().get_group_id()] += r;

  }

  Reduction::T* some_memory;
  UserProvidedLambda f;

};
```

Calling `launch_reduce` for the implementation sketched above would actually launch a sequence of kernels:
1) Initialize `some_memory` to the identity of the reduction.
2) Call the reduction kernel derived from the user-provided lambda.
3) Call some additional kernels to combine all the work-group local results in `some_memory`.

Sorry for all the detail there, but I couldn't think of a shorter way to explain this, and need something to refer back to.

> A backend-specific kernel will never have an argument that is a "reduction" object because the reduction object is a SYCL-specific concept.

That's true, but as you can see above, the SYCL kernel doesn't have any "reduction" object arguments either.  The kernel enqueued by the backend just has a regular `T*` as an argument.  To define backend interop for reduction kernels, a backend would just need to define the expected interface for kernel 2).

> I don't see why we would want a special function name for the case when some arguments are reductions. We would instead expand the set of legal arguments for `handler::set_arg` in general, and then users could use the regular launch overload that takes a kernel object.

I think we could make this work, but I don't think the behavior would be as clear, and it might make things harder to implement.  What you're proposing would require `set_arg` to accept a `sycl::reduction` argument and convert that to a `T*` internally.  The regular `launch` overload taking a kernel object would then need to contain logic to check for reduction arguments, because it would need to conditionally generate calls to kernels 1) and 3).

By splitting the functions out, the regular `launch` can unconditionally assume it's just launching kernels in a straightforward manner, while `launch_reduce` can unconditionally assume it has to call kernels 1) and 3).  My understanding is that this is why @illuhad initially suggested this split, but Aksel please correct me if I'm wrong.

I 100% agree that nobody has defined behavior like this yet.  But since the purpose of this KHR is to gather feedback from users and implementers, I think it makes sense to give backends an opportunity to try and implement reduction interop instead of ruling it out now.

Reply:
@gmlueck and I met and discussed this offline.  He convinced me that there's no reason for this function to exist until we have a situation where a backend actually decides to expose backend interop for reductions.  If that happens, we can always add these definitions back via a revision of the KHR.

I've removed the `launch_reduce` and `launch_grouped_reduce` overloads accepting a kernel object in d26831a0415669dc985a815e3fc89b6f1d34cabb.